src/Common/src/System/Net/CookieParser.cs

```diff
@@ -543,18 +544,60 @@ internal string GetString()
             return _tokenizer.GetCookieString();
         }
 
-        private static void SetCookieName(Cookie cookie, string name)
+        private static MethodInfo s_internalSetNameMethod;
+        private static MethodInfo InternalSetNameMethod
         {
-            try
+            get
             {
-                cookie.Name = name;
+                if (s_internalSetNameMethod == null)
+                {
+                    // We need to use Cookie.InternalSetName instead of the Cookie.set_Name wrapped in a try catch block, as
+                    // Cookie.set_Name keeps the original name if the string is empty or null.
+                    // Unfortunately this API is internal so we use reflection to access it. The method is cached for performance reasons.
+                    MethodInfo method = typeof(Cookie).GetMethod("InternalSetName", BindingFlags.NonPublic | BindingFlags.Instance);
+                    Debug.Assert(method != null, "We need to use an internal method named InternalSetName that is declared on Cookie.");
+                    s_internalSetNameMethod = method;
+                }
+
+                return s_internalSetNameMethod;
```

Comment:
For all of this reflection, it'd be better to cache delegates rather than method infos.